pkg/pilot/cassandra/v3/pilot.go

```diff
@@ -39,9 +45,65 @@ func NewPilot(opts *PilotOptions) (*Pilot, error) {
 		nodeTool:            opts.nodeTool,
 	}
 
+	// hack to test the seedprovider works in k8s, this should be moved somewhere better
+	cfgPath := "/etc/cassandra/cassandra.yaml"
+	read, err := ioutil.ReadFile(cfgPath)
+	if err != nil {
+		return nil, err
+	}
+
+	newContents := strings.Replace(string(read),
+		"org.apache.cassandra.locator.SimpleSeedProvider",
+		"io.k8s.cassandra.KubernetesSeedProvider", -1)
+
+	newContents = strings.Replace(string(newContents),
+		"SimpleSnitch",
+		"GossipingPropertyFileSnitch", -1)
+
+	err = ioutil.WriteFile(cfgPath, []byte(newContents), 0)
+	if err != nil {
+		return nil, err
+	}
+
+	snitchSettings, err := getSnitchSettings(opts.kubeClientset)
+	if err != nil {
+		return nil, err
+	}
+
+	glog.V(4).Info("generated snitch:")
+	glog.V(4).Info(snitchSettings)
+
+	err = ioutil.WriteFile("/cassandra-rackdc.properties", []byte(snitchSettings), 0)
+	if err != nil {
+		return nil, err
+	}
+
 	return p, nil
 }
 
+func getSnitchSettings(clientset kubernetes.Interface) (string, error) {
+	nodeClientset := clientset.CoreV1().Nodes()
+
+	nodeName := os.Getenv("NODE_NAME")
+	node, err := nodeClientset.Get(nodeName, metav1.GetOptions{})
+	if err != nil {
+		return "", err
+	}
+
+	datacenter := "navigator-datacenter-1"
+	rack := "navigator-rack-1"
+
+	if val, ok := node.Annotations["failure-domain.beta.kubernetes.io/region"]; ok {
+		datacenter = val
+	}
+
+	if val, ok := node.Annotations["failure-domain.beta.kubernetes.io/zone"]; ok {
+		rack = val
+	}
```

Comment:
I don't think we can enforce that zone and region are the hardcoded keys we support - there may be some environments where e.g. they want to distribute over many physical RAID arrays, in which case being able to use arbitrary node labels here becomes valuable.

There is a similar feature in Elasticsearch, 'Shard Allocation Awareness' - the only difference with that is it supports more than just DC/Region/Rack (you can add arbitrary 'keys' that can then be reasoned over). I wrote up a small proposal as to the API surface of this here: https://github.com/jetstack/navigator/issues/44#issuecomment-357212909

wdyt? I like the "out of the box it's replicated" behaviour, but should we expose it through a more advanced syntax that has defaults?

Reply:
Realistically we won't be able to make it *that* general purpose (imo). Notably due to the limitations we're already seeing in different DB types.

That said, the use of the two should be *similar*. If someone has been operating Navigator ES databases for some time, they should be able to pick up the Cassandra resource and know how to apply those rules there. (i.e. field names might be different, but field values/types should be similar/the same)